cli/dataset.go

```diff
@@ -150,3 +156,145 @@ func (c *viamClient) deleteDataset(datasetID string) error {
 	return nil
 }
+
+// DatasetDownloadAction is the corresponding action for 'dataset download'.
+func DatasetDownloadAction(c *cli.Context) error {
+	client, err := newViamClient(c)
+	if err != nil {
+		return err
+	}
+	if err := client.downloadDataset(c.Path(dataFlagDestination), c.String(datasetFlagDatasetID),
+		c.Bool(datasetFlagIncludeJSONLines), c.Uint(dataFlagParallelDownloads)); err != nil {
+		return err
+	}
+	return nil
+}
+
+// downloadDataset downloads a dataset with the specified ID.
+func (c *viamClient) downloadDataset(dst, datasetID string, includeJSONLines bool, parallelDownloads uint) error {
+	if err := c.ensureLoggedIn(); err != nil {
+		return err
+	}
+
+	var datasetFile *os.File
+	var err error
+	if includeJSONLines {
+		datasetPath := filepath.Join(dst, "dataset.jsonl")
+		if err := os.MkdirAll(filepath.Dir(datasetPath), 0o700); err != nil {
+			return errors.Wrapf(err, "could not create dataset directory %s", filepath.Dir(datasetPath))
+		}
+		//nolint:gosec
+		datasetFile, err = os.Create(datasetPath)
+		if err != nil {
+			return err
+		}
+		defer func() {
+			if err := datasetFile.Close(); err != nil {
+				Errorf(c.c.App.ErrWriter, "failed to close dataset file %q", datasetFile.Name())
+			}
+		}()
+	}
+
+	return c.performActionOnBinaryDataFromFilter(
+		func(id *datapb.BinaryID) error {
+			downloadErr := downloadBinary(c.c.Context, c.dataClient, dst, id, c.authFlow.httpClient, c.conf.Auth)
+			var datasetErr error
+			if includeJSONLines {
+				datasetErr = binaryDataToJSONLines(c.c.Context, c.dataClient, datasetFile, id)
+			}
+			return multierr.Combine(downloadErr, datasetErr)
+		},
+		&datapb.Filter{
+			DatasetId: datasetID,
+		}, parallelDownloads,
+		func(i int32) {
+			printf(c.c.App.Writer, "Downloaded %d files", i)
+		},
+	)
+}
+
+// Annotation holds the label associated with the image.
+type Annotation struct {
+	AnnotationLabel string `json:"annotation_label"`
+}
+
+// ImageMetadata defines the format of the data in jsonlines for custom training.
+type ImageMetadata struct {
+	ImagePath                 string           `json:"image_path"`
+	ClassificationAnnotations []Annotation     `json:"classification_annotations"`
+	BBoxAnnotations           []BBoxAnnotation `json:"bounding_box_annotations"`
+}
+
+// BBoxAnnotation holds the information associated with each bounding box.
+type BBoxAnnotation struct {
+	AnnotationLabel string  `json:"annotation_label"`
+	XMinNormalized  float64 `json:"x_min_normalized"`
+	XMaxNormalized  float64 `json:"x_max_normalized"`
+	YMinNormalized  float64 `json:"y_min_normalized"`
+	YMaxNormalized  float64 `json:"y_max_normalized"`
+}
+
+func binaryDataToJSONLines(ctx context.Context, client datapb.DataServiceClient, file *os.File,
+	id *datapb.BinaryID,
+) error {
+	var resp *datapb.BinaryDataByIDsResponse
+	var err error
+	for count := 0; count < maxRetryCount; count++ {
+		resp, err = client.BinaryDataByIDs(ctx, &datapb.BinaryDataByIDsRequest{
```

Comment:
What's the reason we're using `ByIDs` but only passing in one ID?

Reply:
Good q, this is mainly to use the batch operations from when we get data from the ByFilter API and then use the IDs to actually get the binary itself [here](https://github.com/viamrobotics/rdk/blob/3fc61c36d9564dc85b5d908c3ebfaaa526c0a27c/cli/dataset.go#L198). I think originally we went 1-1 for simplicity and to avoid timeouts associated with downloading many/larger files.

Reply:
Gotcha! Makes sense! Maybe just add a comment then

Reply:
Let me know if [this](https://github.com/viamrobotics/rdk/blob/3fc61c36d9564dc85b5d908c3ebfaaa526c0a27c/cli/data.go#L287) existing comment makes sense to you or you would like to see something else!

Reply:
Ah, missed this! OK! I'll approve :)